win32ss/printing/base/winspool/printers.c

```diff
@@ -205,9 +205,166 @@ DocumentEvent( HANDLE hPrinter, HDC hdc, int iEsc, ULONG cbIn, PVOID pvIn, ULONG
 LONG WINAPI
 DocumentPropertiesA(HWND hWnd, HANDLE hPrinter, LPSTR pDeviceName, PDEVMODEA pDevModeOutput, PDEVMODEA pDevModeInput, DWORD fMode)
 {
+    PWSTR pwszDeviceName = NULL;
+    PDEVMODEW pdmwInput = NULL;
+    PDEVMODEW pdmwOutput = NULL;
+    BOOL bReturnValue = -1;
+    DWORD cch;
+
     TRACE("DocumentPropertiesA(%p, %p, %s, %p, %p, %lu)\n", hWnd, hPrinter, pDeviceName, pDevModeOutput, pDevModeInput, fMode);
-    UNIMPLEMENTED;
-    return -1;
+
+    if (pDeviceName)
+    {
+        // Convert pName to a Unicode string pwszDeviceName.
+        cch = strlen(pDeviceName);
+
+        pwszDeviceName = HeapAlloc(hProcessHeap, 0, (cch + 1) * sizeof(WCHAR));
+        if (!pwszDeviceName)
+        {
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+            ERR("HeapAlloc failed!\n");
+            goto Cleanup;
+        }
+
+        MultiByteToWideChar(CP_ACP, 0, pDeviceName, -1, pwszDeviceName, cch + 1);
+    }
+
+    if (pDevModeInput)
+    {
+        // Create working buffer for input to DocumentPropertiesW.
+        pdmwInput = HeapAlloc(hProcessHeap, 0, sizeof(DEVMODEW));
+        if (!pdmwInput)
+        {
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+            ERR("HeapAlloc failed!\n");
+            goto Cleanup;
+        }
+
+        MultiByteToWideChar(CP_ACP, 0, (LPCSTR)pDevModeInput->dmDeviceName, -1, pdmwInput->dmDeviceName, CCHDEVICENAME);
```

Comment:
Converting from `DEVMODEA` to `DEVMODEW` and vice-versa is a nontrivial and common operation, so that winspool.drv exports two (undocumented) functions for that:
* `ConvertAnsiDevModeToUnicodeDevmode`
* `ConvertUnicodeDevModeToAnsiDevmode`

Unfortunately, I haven't investigated them yet and don't know of anyone else who did. They are just stubs in our current "winspool.drv".
Exports show that they take 16 bytes of parameters on 32-bit, which probably resolve to InputPointer, InputSize, OutputPointer, OutputSize.
Someone should write a simple test program calling `DocumentPropertiesA` and follow the call chain in rohitab.com's API Monitor. I'm pretty sure this function calls `ConvertAnsiDevModeToUnicodeDevmode` and `ConvertUnicodeDevModeToAnsiDevmode` in the process, so it should be pretty easy to figure out what goes in and what comes out.

Anyway, for now, please refactor your conversion logic into two functions `RosConvertAnsiDevModeToUnicodeDevmode` and `RosConvertUnicodeDevModeToAnsiDevmode` and put them in "devmode.c". Add a `FIXME` comment that these functions shall eventually become `ConvertAnsiDevModeToUnicodeDevmode` and `ConvertUnicodeDevModeToAnsiDevmode` once their parameters have been figured out. Then use both functions from your implementation of `DocumentPropertiesA`.

For the record, there is also `GdiConvertToDevmodeW` in "gdi32.dll", but dependencies show that "winspool.drv" doesn't use that.

Reply:
@ColinFinck, I have made the changes you requested. Please see what you think. Thanks.